src/ray/rpc/raylet/raylet_client.cc

```diff
@@ -0,0 +1,385 @@
+  stub_ = RayletService::NewStub(channel);
+
+  rpc_thread_ = std::thread([this]() { main_service_.run(); });
+  RAY_LOG(DEBUG) << "Connecting to unix socket: "
+                 << "unix://" + raylet_socket
+                 << ", is worker: " << (is_worker_ ? "true" : "false")
+                 << ", worker id: " << worker_id;
+  // Try to register client for 10 times.
+  TryRegisterClient(10);
+}
+
+void RayletClient::TryRegisterClient(int times) {
+  // We should block here until register succeed.
+  for (int i = 0; i < times; i++) {
+    auto st = RegisterClient();
+    if (st.ok()) {
+      Heartbeat();
+      return;
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+  }
+  RAY_LOG(FATAL) << "Failed to register to raylet server, worker id: " << worker_id_
+                 << ", pid: " << static_cast<int>(getpid())
+                 << ", is worker: " << is_worker_;
+}
+
+RayletClient::~RayletClient() {
+  main_service_.stop();
+  rpc_thread_.join();
+}
+
+ray::Status RayletClient::Disconnect() {
+  DisconnectClientRequest disconnect_client_request;
+  disconnect_client_request.set_worker_id(worker_id_.Binary());
+
+  DisconnectClientReply reply;
+  grpc::ClientContext context;
+  auto status = stub_->DisconnectClient(&context, disconnect_client_request, &reply);
+  if (!status.ok()) {
+    RAY_LOG(ERROR) << "Failed to disconnect from raylet, msg: " << status.error_message();
+  }
+  return GrpcStatusToRayStatus(status);
+}
+
+ray::Status RayletClient::SubmitTask(const std::vector<ObjectID> &execution_dependencies,
+                                     const ray::TaskSpecification &task_spec) {
+  SubmitTaskRequest submit_task_request;
+  submit_task_request.set_task_spec(task_spec.Serialize());
+  IdVectorToProtobuf<ObjectID, SubmitTaskRequest>(
+      execution_dependencies, submit_task_request,
+      &SubmitTaskRequest::add_execution_dependencies);
+
+  auto callback = [](const Status &status, const SubmitTaskReply &reply) {
+    if (!status.ok()) {
+      RAY_LOG(INFO) << "Failed to submit task, msg: " << status.message();
+    }
+  };
+
+  auto call =
+      client_call_manager_.CreateCall<RayletService, SubmitTaskRequest, SubmitTaskReply>(
+          *stub_, &RayletService::Stub::PrepareAsyncSubmitTask, submit_task_request,
+          callback);
+  return call->GetStatus();
+}
+
+ray::Status RayletClient::GetTask(std::unique_ptr<ray::TaskSpecification> *task_spec) {
+  GetTaskRequest get_task_request;
+  get_task_request.set_worker_id(worker_id_.Binary());
+
+  grpc::ClientContext context;
+  GetTaskReply reply;
+  // The actual RPC.
+  auto status = stub_->GetTask(&context, get_task_request, &reply);
+
+  resource_ids_.clear();
+  if (status.ok()) {
+    for (size_t i = 0; i < reply.fractional_resource_ids().size(); ++i) {
+      auto const &fractional_resource_ids = reply.fractional_resource_ids()[i];
+      auto &acquired_resources = resource_ids_[fractional_resource_ids.resource_name()];
+
```

Comment:
The following block is a little bit hard to follow - a two-sentence comment here describing what it's doing would go a long way.